text/0002-extracting-necko.md

```diff
@@ -0,0 +1,67 @@
+  - Necko depends on the Mozilla preferences service (libpref) for configuration, which is a privileged singleton that itself depends on the observer service (for startup and profile events) and the directory service (for profile directory and app locations).
+* Because of its dependency on the profile (both for reading state and as a key for the disk cache) and flushing files to disk for prefs, it's not safe to use Necko from two distinct processes simultaneously for the same application.
+* Because of its extensive use of XPCOM singleton patterns, it's not possible to have two independent Neckos in the same process; they need separate processes and separate profile directories. Some independence can be achieved by carefully configuring channels, but it's not isolation.
+* It's not possible to use Necko without a profile service (*e.g.*, to make a network request without touching disk).
+* Parts of Necko are still single-threaded, requiring new channels to be created on the main thread, and dispatching callbacks to the main thread.
+* Necko's init and deinit are managed by observer notifications, which is not ideal for embedding outside of Gecko.
+
+# Proposal
+
+Necko is a very feature-rich C++ networking library that is directly targeted at meeting the needs of single-user, single-main-process, single-profile, interactive applications built on a persistent profile directory in the traditional Mozilla model. It is not well-suited for use outside of the Gecko lifecycle — indeed, almost every aspect is designed for the Gecko world — and would require substantial effort in decoupling to be suited for embedded use. Even if so decoupled, it would still mandate the use of XPCOM and the NSPR.
+
+To extract Necko for reuse outside of Gecko, by non-XPCOM-based applications, would — almost by definition — require coming up with a plan to deal with XPCOM. Given that Necko's data structures, interface definitions, macros, etc. *are* XPCOM's, it seems infeasible to eliminate XPCOM dependency from the module without a total rewrite.
+
+This is particularly relevant when we consider oxidation — the replacement of C++ and JS components of Firefox with Rust — and even more so when we consider the use of such components both outside and inside Firefox *from* Rust (rather than from existing non-Rust callers).
+
+Our position is that extracting Necko from Gecko for reuse would be a significant amount of work, and doing so in a way that allows its use from ordinary Rust consumers, without shipping XPCOM/NSPR/*etc*., would be even more so.
+
+Our proposal, then, is that we should not attempt to reuse Necko outside of Gecko.
```

Comment:
The argument presented over the last two paragraphs sounds like:

1. It would be hard.
2. Therefore we should not do it.

That misses a key step, which is that extracting Necko doesn't provide enough value to justify the work. After all, if it was hard, but it was worth it, then we'd want to do it anyway (c.f. GeckoView, a UA service, etc.).

So I would add a paragraph between "Our position is…" and "Our proposal, then…" that describes the extent of the value we expect to gain from extracting Necko, and how it's not commensurate with the work involved.